WalletWasabi/Wallets/WalletManager.cs

```diff
@@ -0,0 +1,172 @@
+using WalletWasabi.Blockchain.Transactions;
+using WalletWasabi.CoinJoin.Client.Clients.Queuing;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+using WalletWasabi.Services;
+
+namespace WalletWasabi.Wallets
+{
+	public class WalletManager
+	{
+		public WalletManager(string walletBackupsDir)
+		{
+			WalletBackupsDir = walletBackupsDir;
+			Wallets = new Dictionary<WalletService, HashSet<uint256>>();
+			Lock = new object();
+			AddRemoveLock = new AsyncLock();
+		}
+
+		public event EventHandler<ProcessedResult> WalletRelevantTransactionProcessed;
+
+		public event EventHandler<DequeueResult> OnDequeue;
+
+		public Dictionary<WalletService, HashSet<uint256>> Wallets { get; }
+		public object Lock { get; }
+		public AsyncLock AddRemoveLock { get; }
+		public string WalletBackupsDir { get; }
+
+		public IEnumerable<KeyValuePair<WalletService, HashSet<uint256>>> AliveWalletsNoLock => Wallets.Where(x => x.Key is { IsStoppingOrStopped: var isDisposed } && !isDisposed);
+
+		public WalletService GetFirstOrDefaultWallet()
+		{
+			lock (Lock)
+			{
+				return Wallets.Keys.FirstOrDefault();
+			}
+		}
+
+		public async Task AddAndStartAsync(WalletService walletService, CancellationToken cancel)
+		{
+			using (await AddRemoveLock.LockAsync().ConfigureAwait(false))
+			{
+				lock (Lock)
+				{
+					Wallets.Add(walletService, new HashSet<uint256>());
+				}
+
+				Logger.LogInfo($"Starting {nameof(WalletService)}...");
+				await walletService.StartAsync(cancel).ConfigureAwait(false);
+				Logger.LogInfo($"{nameof(WalletService)} started.");
+
+				cancel.ThrowIfCancellationRequested();
+
+				walletService.TransactionProcessor.WalletRelevantTransactionProcessed += TransactionProcessor_WalletRelevantTransactionProcessed;
+				walletService.ChaumianClient.OnDequeue += ChaumianClient_OnDequeue;
+			}
+		}
+
+		private void ChaumianClient_OnDequeue(object sender, DequeueResult e)
+		{
+			var handler = OnDequeue;
+			handler?.Invoke(sender, e);
+		}
+
+		private void TransactionProcessor_WalletRelevantTransactionProcessed(object sender, ProcessedResult e)
+		{
+			var handler = WalletRelevantTransactionProcessed;
+			handler?.Invoke(sender, e);
+		}
+
+		public async Task RemoveAndStopAllAsync()
+		{
+			using (await AddRemoveLock.LockAsync().ConfigureAwait(false))
+			{
+				List<WalletService> walletsListClone;
+				lock (Lock)
+				{
+					walletsListClone = Wallets.Keys.ToList();
+				}
+				foreach (var walletService in walletsListClone)
+				{
+					walletService.TransactionProcessor.WalletRelevantTransactionProcessed -= TransactionProcessor_WalletRelevantTransactionProcessed;
+					walletService.ChaumianClient.OnDequeue -= ChaumianClient_OnDequeue;
+
+					lock (Lock)
+					{
+						if (!Wallets.Remove(walletService))
+						{
+							throw new InvalidOperationException("Wallet service doesn't exist.");
+						}
+					}
+
+					var keyManager = walletService.KeyManager;
+					if (keyManager is { } && WalletBackupsDir is { })
```

Comment:
I think we should be using `!string.IsNullOrWhiteSpace ()` on string types?

Reply:
Done.